zeus/metric.py

```diff
@@ -0,0 +1,477 @@
+    ) -> None:
+        """
+        Initialize the EnergyHistogram class.
+
+        Sets up the Prometheus Histogram metrics to track energy consumption for GPUs, CPUs, and DRAMs.
+        The data will be collected and pushed to the Prometheus Push Gateway at regular intervals.
+
+        Args:
+            energy_monitor: The ZeusMonitor instance used to retrieve energy data for the system.
+            prometheus_url: The URL for the Prometheus Push Gateway where the metrics will be sent.
+            job: The name of the Prometheus job associated with the energy metrics.
+            bucket_ranges: Optional custom bucket ranges for the Histogram metrics (GPU, CPU, and DRAM).
+                        If not provided, default bucket ranges will be used for each component.
+        """
+        self.energy_monitor = energy_monitor
+        self.prometheus_url = prometheus_url
+        self.job = job
+
+        self.registry = CollectorRegistry()
+
+        default_gpu_buckets = [50.0, 100.0, 200.0, 500.0, 1000.0]
+        default_cpu_buckets = [10.0, 20.0, 50.0, 100.0, 200.0]
+        default_dram_buckets = [5.0, 10.0, 20.0, 50.0, 150.0]
+
+        self.bucket_ranges = {
+            "gpu": default_gpu_buckets,
+            "cpu": default_cpu_buckets,
+            "dram": default_dram_buckets,
+        }
+
+        self.bucket_ranges["gpu"] = (
+            bucket_ranges.get("gpu")
+            if bucket_ranges and "gpu" in bucket_ranges
+            else default_gpu_buckets
+        )
+
+        self.bucket_ranges["cpu"] = (
+            bucket_ranges.get("cpu")
+            if bucket_ranges and "cpu" in bucket_ranges
+            else default_cpu_buckets
+        )
+
+        self.bucket_ranges["dram"] = (
+            bucket_ranges.get("dram")
+            if bucket_ranges and "dram" in bucket_ranges
+            else default_dram_buckets
+        )
+        # If GPU availble, for each gpu_indices, create a Histogram metric with the label window, and index.
+        if energy_monitor.gpu_indices:
+            self.gpu_histograms = {}
+            for gpu_index in self.energy_monitor.gpu_indices:
+                self.gpu_histograms[gpu_index] = Histogram(
+                    f"energy_monitor_gpu_{gpu_index}_energy_joules",
+                    f"GPU {gpu_index} energy consumption",
+                    ["window", "index"],
+                    buckets=self.bucket_ranges.get("gpu", []),
+                    registry=self.registry,
+                )
+        else:
+            self.gpu_histogram = None
+        # If CPU available, for each cpu_indices, create a Histogram metric with the label window, and index.
+        if energy_monitor.cpu_indices:
+            self.cpu_histograms = {}
+            for cpu_index in self.energy_monitor.cpu_indices:
+                self.cpu_histograms[cpu_index] = Histogram(
+                    f"energy_monitor_cpu_{cpu_index}_energy_joules",
+                    f"CPU {cpu_index} energy consumption",
```

Comment:
Question: Are these the name of the metric and its description?

If so, I think these should also be documented. That is, given cpu indices and window names, how the exact name of the metrics from our metric classes are constructed.

Also, do you think the user might want to override this based on their needs? I think I currently have a poor understanding of how metric names are organized in Prometheus. How are these two parameters, `job`, and the window name organized from the user's perspective, looking at their Prometheus dashboard?

All these ought to be documented. Instead of answering this comment, you can just update the documentation and let me know.